### PR TITLE
Add thread checks to emscripten::val

### PR DESCRIPTION
**Pull request by RReverser**
In multithreading environment using `emscripten::val` is very error-prone and hard to debug. The reason for that is that, on one hand, `val` represents JavaScript values which cannot be (easily) shared between threads, but, on another, `val` itself is a thin wrapper around an integer handle which can be shared across threads way too easily.

As a result, accessing `val` from the wrong thread sometimes succeeds by accident (such as for built-in constants like `undefined`/`false`/`true`), sometimes crashes with invalid handle, and sometimes just quietly returns completely different JS value because a handle with the same ID happened to be valid on both threads, but points to different underlying values.

In the last scenario code can even keep going forward for some time, using and making changes on the incorrect JS value before it's apparent that something went very wrong.

To make it easier to catch such cases, in this PR I'm adding a `pthread_t` field to the `val` class that stores the owner thread of the given handle. User can still store `val` and send it across threads, but the moment they try to operate on the underlying JS value on the wrong thread, an assertion will fire with a clear message.

This check is activated only when code is compiled with pthread support, and only if user didn't disable assertions via `NDEBUG` (we can't use the `-sASSERTIONS` setting here because it's a header not compiled code). This still adds a tiny size overhead if the user is not passing `-DNDEBUG`, but typical build systems (e.g. CMake) pass it automatically in release builds, and overall I believe benefits outweigh the overhead in this scenario.

In the future we might also automatically proxy all operations to the correct thread so that `emscripten::val` behaves like any other native object. However, that's out of scope of this particular PR, so for now just laying the groundwork and making those failure scenarios more visible.

### CODE REVIEW

File: system/include/emscripten/val.h
Comment (sbc100) [inline system/include/emscripten/val.h:636]: Why the underscore suffix?  If that is for private members why not also add that for `thread` above?

File: system/include/emscripten/val.h
Comment (sbc100) [inline system/include/emscripten/val.h:431]: Are the destructor/copy/assignment changes here a separate cleanup/fix?  Can/should then land separately/first?

File: system/include/emscripten/val.h
Comment (sbc100) [inline system/include/emscripten/val.h:411]: Should this use `as_handle` (same with the destructor below)?

Comment (RReverser) [reply system/include/emscripten/val.h:411]: Yeah just noticed those too and updating. Thanks!

File: system/include/emscripten/val.h
Comment (RReverser) [reply system/include/emscripten/val.h:431]: I can split them out, but they were useful here in particular because handle assignments got more complicated in this PR.

File: system/include/emscripten/val.h
Comment (RReverser) [reply system/include/emscripten/val.h:636]: TBH it was mostly for me to quickly notice error squiggles everywhere `handle` is used, but then decided to leave as-is as sort of a "private" notation, yeah. I'm happy to either rename back or rename `thread` too.

File: system/include/emscripten/val.h
Comment (RReverser) [reply system/include/emscripten/val.h:403]: Pushed.

File: system/include/emscripten/val.h
Comment (sbc100) [reply system/include/emscripten/val.h:628]: Maybe lets just leave it without the underscore then?

Comment (RReverser) [reply system/include/emscripten/val.h]: Renamed back.

File: test/test_core.py
Comment (sbc100) [inline test/test_core.py:7779]: I think we have a `@node_pthreads` decorator for this

**Review by sbc100 — APPROVED** [review]
lgtm % comments

This is a great idea BTW!

File: test/test_core.py
Comment (RReverser) [reply test/test_core.py:7779]: Ah right.

**Comment by RReverser** [comment]
One small concern I have is that this is technically a breaking change.

In particular, it was possible "reserved" values like `undefined`/`false`/`true`/`null` which have same IDs on all threads would work with cross-thread access before, but are forbidden now.

Is it worth adding extra condition to allow those special values to be access from any thread, or is it better to keep things strict and uniform for all values?

**Comment by sbc100** [comment]
> One small concern I have is that this is technically a breaking change.
> 
> In particular, "reserved" values like `undefined`/`false`/`true`/`null` which have same IDs on all threads would work with cross-thread access before, but are forbidden now.
> 
> Is it worth adding extra condition to allow those special values to be accessed from any thread, or is it better to keep things strict and uniform for all values?

I don't think that is worth it no.   I don't think sharing those values is a good idea.  Perhaps just an a ChangeLog entry and if folks out there really were depending on sharing just those core values we can consider adding that extra allowance.

@brendandahl WDYT?

**Comment by RReverser** [comment]
> I don't think sharing those values is a good idea.

I agree, it's safer to forbid any kind of cross-thread access, especially given that it's implementation detail. Just worried about fallout if any.

> Perhaps just an a ChangeLog entry

Oh right, this definitely warrants a changelog entry which I forgot about.

**Comment by brendandahl** [comment]
I think it's good to forbid it for all vals. This has a been a pretty common issue that I've seen people run into with pthreads+embind.

File: system/include/emscripten/val.h
Comment (RReverser) [reply system/include/emscripten/val.h:423]: Should I split them out or ok to leave in this PR?

Comment (sbc100) [reply system/include/emscripten/val.h:423]: I don't feel strongly.    I always prefer smaller PR and I have C++ move/copy operators a bit mysterious, but as you point out maybe this change isn't very useful when there is just a single member in the class ?   If you think its an independently useful change that I would split it out.

Comment (RReverser) [reply system/include/emscripten/val.h:423]: I think it could make PR easier to review if that was a problem but otherwise it's only useful in context of this PR, yeah.

**Comment by sbc100** [comment]
One potential pitfall there is that the `sizeof(val)` will be different in compilation units that define `NDEBUG` and ones that don't.   Do unless all object files and libraries in the entire project are compiled similarly this could lead to bad things.  

How many val's are there in most programs?  Would unconditionally adding the new field for mutli-threaded applications have a significant impact?

File: system/include/emscripten/val.h
Comment (sbc100) [reply system/include/emscripten/val.h:423]: Is this particular 3 line pattern a common way to implement rvalue assignment?  It looks very generic .. is it different the `= default` version?  (Showing my C++ ignorance here).

**Comment by RReverser** [comment]
> Do unless all object files and libraries in the entire project are compiled similarly this could lead to bad things.

Hm yeah that's a good point.

> How many val's are there in most programs? Would unconditionally adding the new field for mutli-threaded applications have a significant impact?

I was mostly concerned about something like `std::vector<val>`, but either way I don't think it will be a problem in any real-world app.

As mentioned in PR description, we might want to add it unconditionally anyway in the future.

To be clear, `_REENTRANT` is still fine though? I don't remember if we allow linking objects compiled with and without pthread nowadays.

File: system/include/emscripten/val.h
Comment (sbc100) [reply system/include/emscripten/val.h:423]: This seems like a doses a bunch of extra incrementing a decrementing of the counter.   I assume tmp here to keep the refcount above zero when the descructor runs?   Why bother doing that?

Comment (RReverser) [reply system/include/emscripten/val.h:423]: > Is this particular 3 line pattern a common way to implement rvalue assignment? It looks very generic .. is it different the = default version? (Showing my C++ ignorance here).

I'm not sure either... I came up with this particular sequence on my own.

> I assume tmp here to keep the refcount above zero when the descructor runs? Why bother doing that?

Because there is a test that verifies we can do `val = std::move(val);` or something like that (assign value by move to itself) which failed before.

Comment (brendandahl) [reply system/include/emscripten/val.h:423]: More context in https://github.com/emscripten-core/emscripten/pull/20072

**Comment by RReverser** [comment]
Removed most ifdefs; kept only one around assertion itself - it seems reasonable that if someone compiles without `-pthread`, they don't expect their values to be shared across threads, and this way single-threaded version of Embind remains cheap.

Of course, `NDEBUG` also still applies since that's part of `assert` macro itself.

**Comment by RReverser** [comment]
Odd, the failing test (`wasm64_4gb.test_embind_val`) seems clearly relevant, yet it passes fine for me locally with Node.js 20.7.0.

**Review by sbc100 — APPROVED** [review]

**Comment by RReverser** [comment]
Looks like that particular test is failing pretty consistently on CI. Could use some help there as can't reproduce.

Could it be due to CI using newer / canary Node.js than the latest stable?

**Comment by RReverser** [comment]
Hm no, downloaded that newer Node.js and can't reproduce with it either. So odd.

**Comment by RReverser** [comment]
Looks like this somehow broke `size_t` index conversion so it's now getting wrapped around to unsigned uint64 before being passed to `emval_get_property`.

```
test: template<typename T> std::vector<T> vecFromJSArray(const val& v)...
{
  handle: [ 1, '42', 'b', 1970-01-01T00:01:40.000Z ],
  key: 'length',
  value: 4
}
{
  handle: [ 1, '42', 'b', 1970-01-01T00:01:40.000Z ],
  key: 0n,
  value: 1
}
{
  handle: [ 1, '42', 'b', 1970-01-01T00:01:40.000Z ],
  key: 4294967297n,
  value: undefined
}
{
  handle: [ 1, '42', 'b', 1970-01-01T00:01:40.000Z ],
  key: 4294967298n,
  value: undefined
}
{
  handle: [ 1, '42', 'b', 1970-01-01T00:01:40.000Z ],
  key: 4294967299n,
  value: undefined
}
{
  handle: [ 1, '42', 'b', 1970-01-01T00:01:40.000Z ],
  key: 4294967297n,
  value: undefined
}
```

**Comment by RReverser** [comment]
Raised separately in https://github.com/emscripten-core/emscripten/pull/20353. I'm guessing this PR is blocked until that's resolved.

**Comment by RReverser** [comment]
Added a workaround just for that method to unblock this.

**Comment by RReverser** [comment]
Heh having some fun fallout from this in one of my projects already where I was doing something that looks safe-ish like

```cpp
val valueStorage;

proxyToMainThread([] {
  valueStorage = someValue();
});

// later `valueStorage` is correctly freed also on main thread
```

Probably still right decision, as this arguably should be `std::optional`, but I imagine it might break few other people as well.

**Comment by RReverser** [comment]
Oh damn, there is also a bug here. Fixing.

**Comment by gmarella** [comment]
@RReverser 
I am hitting the assert added for the thread when I use mimalloc(or tcmalloc). I am currently debugging this, but wanted to check if there is any known issue when using custom allocators like mimalloc with these safeguard checks.

**Pull request merged**
